provider/cmd/leaseEvents.go

```diff
@@ -36,54 +40,84 @@ func doLeaseEvents(cmd *cobra.Command) error {
 		return err
 	}
 
-	prov, err := providerFromFlags(cmd.Flags())
+	cert, err := cutils.LoadAndQueryCertificateForAccount(cmd.Context(), cctx, cctx.Keyring)
 	if err != nil {
 		return err
 	}
 
-	bid, err := mcli.BidIDFromFlagsForOwner(cmd.Flags(), cctx.FromAddress)
+	dseq, err := dseqFromFlags(cmd.Flags())
 	if err != nil {
 		return err
 	}
 
-	svcs, err := cmd.Flags().GetString(FlagService)
+	leases, err := leasesForDeployment(cmd.Context(), cctx, cmd.Flags(), dtypes.DeploymentID{
+		Owner: cctx.GetFromAddress().String(),
+		DSeq:  dseq,
+	})
 	if err != nil {
 		return err
 	}
 
-	follow, err := cmd.Flags().GetBool("follow")
+	svcs, err := cmd.Flags().GetString(FlagService)
 	if err != nil {
 		return err
 	}
 
-	cert, err := cutils.LoadAndQueryCertificateForAccount(cmd.Context(), cctx, cctx.Keyring)
+	follow, err := cmd.Flags().GetBool("follow")
 	if err != nil {
 		return err
 	}
 
-	gclient, err := gwrest.NewClient(akashclient.NewQueryClientFromCtx(cctx), prov, []tls.Certificate{cert})
-	if err != nil {
-		return err
-	}
+	ctx := cmd.Context()
 
-	result, err := gclient.LeaseEvents(cmd.Context(), bid.LeaseID(), svcs, follow)
-	if err != nil {
-		return showErrorToUser(err)
+	type result struct {
+		lid      mtypes.LeaseID
+		provider sdk.Address
+		error    error
+		stream   *gwrest.LeaseKubeEvents
 	}
 
-	for res := range result.Stream {
-		if err = cmdcommon.PrintJSON(cctx, res); err != nil {
-			break
+	streams := make([]result, 0, len(leases))
+
+	for _, lid := range leases {
+		stream := result{lid: lid}
+		prov, _ := sdk.AccAddressFromBech32(lid.Provider)
+		gclient, err := gwrest.NewClient(akashclient.NewQueryClientFromCtx(cctx), prov, []tls.Certificate{cert})
+		if err == nil {
+			stream.stream, stream.error = gclient.LeaseEvents(ctx, lid, svcs, follow)
+		} else {
+			stream.error = err
 		}
+
+		streams = append(streams, stream)
 	}
 
-	select {
-	case msg, ok := <-result.OnClose:
-		if ok && msg != "" {
-			_ = cctx.PrintString(msg + "\n")
+	var wgStreams sync.WaitGroup
+
+	for _, stream := range streams {
+		if stream.error == nil {
+			wgStreams.Add(1)
+			go func(stream result) {
+				defer wgStreams.Done()
+
+				for res := range stream.stream.Stream {
+					type logEntry struct {
+						cltypes.LeaseEvent `json:",inline"`
+						Lid                mtypes.LeaseID `json:"lease_id"`
+					}
+
+					evt := &logEntry{
+						LeaseEvent: res,
+						Lid:        stream.lid,
+					}
+
+					_ = cmdcommon.PrintJSON(cctx, evt)
```

Comment:
does `cmdcommon.PrintJSON` have a mutex?  If not I think that different prints may be interleaved together.

you could change this to

```go
outch <- evt
```

and have something like:

```go
go func(){
  for evt := range outch {
    _cmdcommon.PrintJSON(cctx,evt)
  }
}()
```

and

```go
wgstreams.Wait()
close(outch)
```

Reply:
good point, stdout (default stream) does not guarantee threadsafety